### ISSUE TITLE
Dynamics 365 CRM integration improvements and feedback

### ISSUE THREAD

**Issue by jamesmacwhite** [verbb/formie#1262]
### What are you trying to do?

So I do apologise for spamming the issue queue regarding Dynamics 365 CRM lately. As the organisation I work for has a fairly large and customised CRM we have been finding some interesting areas and quirks with the base integration, but after your suggestion of extending the original class and making modifications we are able to leverage the original with additional enhancement which is working well.

There are some areas which I think would benefit everyone, some may be a bit too bespoke for our needs, but possibly there's general ways to allow the same behaviour we have done in our extended integration, by either further events or just modifying the base integration class. I'll note some of the main areas for further discussion, I would be happy to PR them, but thought I should document them first and maybe get any other Dynamics 365 CRM people commenting

### Provide boolean true/false default values in the mapping for any boolean type fields https://github.com/verbb/formie/issues/1250

This can be added fairly painlessly without an additional API request, basically looping through all the fields after the `new IntegrationField()` call, as the type has been converted at this point and simply check for any boolean fields. You can then insert custom options on these fields, which in this case simply true or false.

While you can query the boolean metadata for each entity, it's not really needed and wasteful as you've defined the type on the field being initialised through the IntegrationField already.

```php
// In a for loop for all the fields in the mapping, check for booleans
if ($field->type === IntegrationField::TYPE_BOOLEAN) {
    $field->options = [
          'label' => Craft::t('formie', 'Default options'),
          'options' => [
	          ['label' => Craft::t('formie', 'True'), 'value' => 'true'],
	          ['label' => Craft::t('formie', 'False'), 'value' => 'false']
          ]
   ];
}
```

Now any boolean field has a true/false value able to be mapped without needing a specific Formie field.

### Add an event to insert additional $targetSchemas items/merge with existing defaults

The target schemas populates some lookup fields but won't populate custom fields that are outside base entities. To allow further lookup population in the mapping on more exotic fields, an event here to insert additional array items would be beneficial to achieve this. An additional custom entity might looks something like this:

```php
'ccl1000_enquirytype' => [
    'entity' => 'ccl1000_enquirytypes',
    'label' => 'ccl1000_name',
    'value' => 'ccl1000_enquirytypeid'
]
```

### Allow further query parameters in the Lookup Guzzle query e.g. orderby or filter

For some lookup fields it populates the first 100 values which may not be in a logical order for that field. Example, campaign name would benefit from being ordered by newest to oldest potentially. This can be done by orderby, filtering is useful when the default request returns values you might not want. Setting a limit is important for performance of course, but being able to override it per entity might be nice.

The Guzzle request can specify additional query parameters but fallback to null if not present in any entity and this will automatically not include it in the request.

e.g.

```php
$response = $this->request('GET', $targetSchema['entity'], [
	'query' => [
		'$top' => $targetSchema['limit'] ?? '100',
		'$select' => implode(',', $select),
		'$orderby' => $targetSchema['orderby'] ?? null,
		'$filter' => $targetSchema['filter'] ?? null
	]
]);
```

Such a modification shouldn't break anything if one or more of the query parameter don't exist for that entity, given using `?? null` will just remove it from the Guzzle request.

Combined with an event that allows you to add further entity items, this would allow for further customisation of the mapping values where needed.

An example of a filter could be for a specific entity that's been picked up but is pulling data that's not needed. Taking a standard entity like systemusers, our default is not ideal, we actually have several thousand systemusers without any filtering. While the query is limited to 100 by default (and that's a good thing!) I passed in an extra filter like:

```
isdisabled eq false and invitestatuscode eq 4 or accessmode eq 4
```

This basically returns active CRM users and any service/non-interactive accounts. The logic where a systemuser exists with an applicationid further excludes some non required users.

### Add a small function to format values which require the `entitypluralname(GUID)` format,

There's a lot of a duplication of string concatenation repeating this format, instead a function would do it nicely, saving you having to write it out each time, as it's a little awkward with string concat multiple times.

```php
private function _formatLookupValue($entity, $value): string
{
    return $entity . '(' . $value . ')';
}
```

Then any use of this would simply be `$this->_formatLookupValue('contacts', $guid)`;

### Assigning contactid, leadid, accountid or opportunityid to additional relational fields in payloads

So this one is a bit trickier, but when each payload is executed, the id of that entity is stored for relational linking on the standard relation fields on entities. In several cases we have different relational entities outside of the standard ones, which require population, like this example.

```php
$leadPayload['parentcontactid@odata.bind'] = $this->_formatLookupValue('contact', $contactId);
// An additional relational field
$leadPayload['customerid_contact@odata.bind'] = $this->_formatLookupValue('contact', $contactId);
```

Doing this through events might be tricky, given you basically need to intercept the payload before it's sent, but also pass the potentially different entity ids across each one, but before it's been sent to the API.

### What's your proposed solution?

Happy to PR some of these, if you think they are worthwhile!

**Comment by engram-design** [comment]
Thanks for the write-up!

### Provide boolean true/false default values in the mapping for any boolean type fields 
Happy for this to be a PR, makes total sense. We'll probably keep this at the individual integration level, and not the overall `Integration` class level, as I know some integrations like to have `Yes/No` values for their boolean fields.

### Add an event to insert additional $targetSchemas items/merge with existing defaults
Fair enough!

### Allow further query parameters in the Lookup Guzzle query e.g. orderby or filter
Sounds good to me.

### Add a small function to format values which require the entitypluralname(GUID) format,
Agreed.

### Assigning contactid, leadid, accountid or opportunityid to additional relational fields in payloads
I think that makes sense to make it part of core, we're _supposed_ to be doing that already [here](https://github.com/verbb/formie/blob/70553023cb2a2e6a57793d569f2603908075aec7/src/integrations/crm/MicrosoftDynamics365.php#L250) so looks like it's just a matter of adding some more relations, so long as they are part of the core object, and not a custom field you add on your end.

**Comment by jamesmacwhite** [comment]
Awesome. I'll look at some PRs in the coming days.

For the $targetSchemas something like this should work:

```php

$targetSchemas = []; // Base ones in the integration class

$event = new MicrosoftDynamics365TargetSchemas([
	'targetSchemas' => $targetSchemas,
]);

$this->trigger(self::EVENT_MODIFY_TARGET_SCHEMAS, $event);

$targetSchemas = ArrayHelper::merge($targetSchemas, $event->targetSchemas);
```

Then in a module you could do:

```php
Event::on(
	MicrosoftDynamics365::class,
	MicrosoftDynamics365::EVENT_MODIFY_TARGET_SCHEMAS,
	static function(MicrosoftDynamics365TargetSchemas $event) {
		$event->targetSchemas = [
			'systemuser' => [
				// Extra filter to pull active users and service accounts
				'filter' => "isdisabled eq false and invitestatuscode eq 4 or accessmode eq 4",
			],
                        // Modify original campaign entity with orderby and limit
			'campaign' => [
				'orderby' => 'createdon desc',
				'limit' => '150',
			],
                       // Custom entity to include on relational fields fetch
			'ccl1000_enquirytype' => [
				'entity' => 'ccl1000_enquirytypes',
				'label' => 'ccl1000_name',
				'value' => 'ccl1000_enquirytypeid'
			]
		];
	}
);
```

On the guzzle request, include the limit, filter and orderby keys and fallback to null if not provided by each entity.

Seems like the cleanest way to do it without modifying the targetSchemas in the base integration class itself. You can keep the original targetSchemas as is as a out of the box standard. Doing an array merge seems like the best way, so you can override existing ones while also adding new entities if required. You also wouldn't need to redefine any of the existing keys present in the default array, just modifications you need.

**Comment by jamesmacwhite** [comment]
Implemented in #1263
